redbot/cogs/mutes/mutes.py

```diff
@@ -0,0 +1,1219 @@
+import asyncio
+import contextlib
+import discord
+import logging
+
+from abc import ABC
+from copy import copy
+from typing import cast, Optional, Dict, List, Tuple, Literal, Coroutine, Union
+from datetime import datetime, timedelta, timezone
+
+from .converters import MuteTime
+from .voicemutes import VoiceMutes
+
+from redbot.core.bot import Red
+from redbot.core import commands, checks, i18n, modlog, Config
+from redbot.core.utils.chat_formatting import humanize_timedelta, humanize_list, pagify
+from redbot.core.utils.mod import get_audit_reason
+from redbot.core.utils.menus import start_adding_reactions
+from redbot.core.utils.predicates import MessagePredicate, ReactionPredicate
+
+T_ = i18n.Translator("Mutes", __file__)
+
+_ = lambda s: s
+mute_unmute_issues = {
+    "already_muted": _("That user is already muted in this channel."),
+    "already_unmuted": _("That user is not muted in this channel."),
+    "hierarchy_problem": _(
+        "I cannot let you do that. You are not higher than the user in the role hierarchy."
+    ),
+    "is_admin": _("That user cannot be muted, as they have the Administrator permission."),
+    "permissions_issue": _(
+        "Failed to mute or unmute user. I need the manage roles "
+        "permission and the user I'm muting must be "
+        "lower than myself in the role hierarchy."
+    ),
+    "left_guild": _("The user has left the server while applying an overwrite."),
+    "unknown_channel": _("The channel I tried to mute the user in isn't found."),
+    "role_missing": _("The mute role no longer exists."),
+}
+_ = T_
+
+log = logging.getLogger("red.cogs.mutes")
+
+__version__ = "1.0.0"
+
+
+class CompositeMetaClass(type(commands.Cog), type(ABC)):
+    """
+    This allows the metaclass used for proper type detection to
+    coexist with discord.py's metaclass
+    """
+
+    pass
+
+
+class Mutes(VoiceMutes, commands.Cog, metaclass=CompositeMetaClass):
+    """
+    Stuff for mutes goes here
+    """
+
+    def __init__(self, bot: Red):
+        self.bot = bot
+        self.config = Config.get_conf(self, 49615220001, force_registration=True)
+        default_guild = {
+            "sent_instructions": False,
+            "mute_role": None,
+            "respect_hierarchy": True,
+            "muted_users": {},
+            "default_time": {},
+            "removed_users": {},
+        }
+        self.config.register_guild(**default_guild)
+        self.config.register_member(perms_cache={})
+        self.config.register_channel(muted_users={})
+        self._server_mutes: Dict[int, Dict[int, dict]] = {}
+        self._channel_mutes: Dict[int, Dict[int, dict]] = {}
+        self._ready = asyncio.Event()
+        self.bot.loop.create_task(self.initialize())
+        self._unmute_tasks: Dict[str, Coroutine] = {}
+        self._unmute_task = asyncio.create_task(self._handle_automatic_unmute())
+        # dict of guild id, member id and time to be unmuted
+        self.mute_role_cache = {}
```

Comment:
(optional) It could potentially help a future reader, if this had a type hint:
```suggestion
        self.mute_role_cache: Dict[int, int] = {}
```